### PR TITLE
[WS2][PR2][Logp] feat: add TP=1 logprob comparison harness

### PR DESCRIPTION
**Pull request by hihaluemen**
## Summary

Add the TP=1 logprob comparison harness requested by PR2 of #241.

The harness registers the existing WS1 batch-invariant PyTorch logprob path as the reference and compares the supported single-GPU backends against it before tensor-parallel communication is introduced. It reports direct vocabulary-LSE drift and active-token-only selected-logprob drift, while recording enough backend provenance to detect accidental fallback.

The existing production operator contract remains unchanged. The new `forward_with_lse` methods are diagnostic entry points used by the comparison harness and tests.

Implements PR2 of #241.

## Scope

This PR covers the single-GPU registration and regression guard described in PR2:

- Use the merged WS1 batch-invariant PyTorch implementation as the reference.
- Require the TP=1 PyTorch path to remain bitwise equal to that reference.
- Compare the PyTorch, Triton, and CUDA SM90 batch-invariant logprob backends.
- Report vocabulary-LSE drift over all logical token rows.
- Report selected-token dlogp drift over active response/action tokens only.
- Record `tp_world=1` and `communication=none` in the report.
- Fail closed when an explicitly requested backend is unavailable or falls back to another implementation.

This PR does not implement vocab sharding, collective communication, fixed-order cross-rank LSE merging, CP reconstruction, or distributed artifact generation. Those remain part of the later PRs in #241.

### Single-GPU comparison harness

Add `rl_engine/testing/logprob_comparison.py` with:

- Structured comparison inputs, candidates, reports, and backend provenance.
- Exact backend selection for `pytorch`, `triton`, and `cuda-sm90`.
- Direct comparison against the existing batch-invariant PyTorch reference.
- Bitwise equality reporting for selected logprobs.
- Max, mean, p95, and p99 absolute-drift statistics.
- Active-token masking for selected-logprob drift.
- Validation for target shape, dtype, range, and `ignore_index` usage.
- A generic operator-comparison registration for `batch_invariant_logp`.

### Diagnostic LSE entry points

Add a diagnostic-only method to each supported backend:

```python
op.forward_with_lse(logits, target_ids, ignore_index=-100) -> (logp, lse)
```

The normal production call remains:

```python
op(logits, target_ids, ignore_index=-100) -> logp
```

The diagnostic path exposes the LSE computed by the backend itself. The harness does not reconstruct LSE from selected logprobs, which keeps the LSE comparison independent and useful for later TP work.

For an explicit `cuda-sm90` request, the diagnostic path requires the compiled SM90 extension and compatible Hopper inputs. It does not use the production operator's fallback behavior.

### Kernel-local command-line entry

Make `rl_engine/testing/logprob_comparison.py` directly executable for reproducible local
and GPU comparisons. This keeps the kernel-specific diagnostic next to its harness instead
of adding a project-wide script.

Example:

```bash
python rl_engine/testing/logprob_comparison.py \
  --candidate triton \
  --candidate cuda-sm90 \
  --device cuda \
  --dtype bf16 \
  --batch 2 \
  --seq 16 \
  --vocab 151936
```

The command writes a structured JSON report to stdout. RL-Kernel diagnostic logs are routed to stderr so redirected stdout remains valid machine-readable JSON.

### Operator documentation

Document the TP=1 harness contract, exact backend selection, CLI usage, and SM90 validation
in `docs/operators/batch-invariant-logp.md`. The content is consolidated into the existing
operator documentation rather than maintained as separate WS2 design documents.

## Comparison contract

For each logical token row, the compared values are:

```text
LSE  = logsumexp(logits[..., vocab])
logp = selected_logit - LSE
```

LSE drift is measured over every logical token row. Selected-logprob drift is measured only where the active-token mask is true. Each drift report contains:

```text
active_count
max_abs
mean_abs
p95_abs
p99_abs
```

The report also records:

- Requested and actual backend.
- Concrete implementation class.
- Direct-LSE provenance.
- Input shape and dtype.
- Active-token count.
- TP world size.
- Communication mode.
- Bitwise selected-logprob status.

## Tests

Add focused coverage for:

- TP=1 PyTorch bitwise regression.
- Direct LSE identity.
- Active-token-only percentile calculation.
- The zero-active-token case.
- Invalid active `ignore_index` usage.
- Structured report serialization.
- Operator-harness registration.
- Exact Triton and CUDA SM90 diagnostic paths.
- Fail-closed backend provenance.
- Machine-readable CLI stdout when RL-Kernel emits diagnostic logs.

### Windows CPU

```bash
python -m pytest tests/test_logprob_comparison.py tests/test_operator_inputs.py tests/test_op_checks.py -q
```

Result:

```text
39 passed, 2 skipped
```

The skipped cases require CUDA/Triton backends.

### WSL

Current focused and complete operator validation:

```text
PR2 focused tests: 42 passed, 1 skipped
Complete batch-invariant logprob suite: 53 passed, 14 skipped
```

The skipped cases require a compiled CUDA SM90 extension on Hopper hardware. The focused
suite includes a subprocess test of the direct module CLI and machine-readable JSON output.

### NVIDIA H800 / SM90

Validated on:

```text
GPU: NVIDIA H800 PCIe
Compute capability: 9.0
Python: 3.11.15
PyTorch: 2.11.0+cu128
CUDA toolkit / nvcc: 12.8
Triton: 3.6.0
```

The editable CUDA extension built successfully with the SM90 kernel enabled:

```text
batch_invariant_logp_sm90=True
```

Test results:

```text
PR2 focused tests: 41 passed in 3.25s
Complete batch-invariant logprob suite: 67 passed in 4.60s
```

Observed BF16 SM90 drift against the PyTorch reference:

| Shape | LSE max abs | dlogp max abs |
| --- | ---: | ---: |
| `[2, 8, 1024]` | `4.76837158203125e-07` | `4.76837158203125e-07` |
| `[2, 16, 151936]` | `9.5367431640625e-07` | `9.5367431640625e-07` |

Both comparisons used `tp_world=1`, `communication=none`, and the requested `cuda-sm90` implementation without fallback.

Additional checks:

```text
MyPy: no issues found in 90 source files
pre-commit --all-files: passed
Python compileall: passed
git diff --check: passed
```

## Notes for review

- The main addition is the comparison and reporting harness; this PR does not change distributed logprob mathematics.
- Production calls still return only selected logprobs.
- `forward_with_lse` exists to expose backend-native diagnostics without changing production callers.
- Explicit backend requests intentionally fail instead of silently falling back, because backend provenance is part of the regression contract.
- The PyTorch bitwise check is the TP=1 regression guard requested by #241; numerical drift is expected for independently implemented Triton and CUDA reductions.

## Summary by CodeRabbit

* **New Features**
  * Added access to both selected log-probabilities and row-wise log-sum-exp values across supported PyTorch, Triton, and SM90 CUDA implementations.
  * Added a single-GPU comparison tool for measuring numerical drift between log-probability backends.
  * Added structured comparison reports, provenance details, and command-line output.

* **Documentation**
  * Added usage, validation, build, and testing guidance for the comparison workflow.

* **Tests**
  * Expanded coverage for backend validation, diagnostics, drift reporting, and command-line behavior.

### CODE REVIEW

File: docs/design/ws2-logprob-single-gpu-harness.md
Comment (KJLdefeated) [inline docs/design/ws2-logprob-single-gpu-harness.md:1]: To let future maintenance easier, I think it is not necessary to create two docs for this. I suggest that you can compress the context of logprob harness and put into existing docs/operators/batch-invariant-logp.md.

File: docs/design/ws2-logprob-sm90-validation.md
Comment (KJLdefeated) [inline docs/design/ws2-logprob-sm90-validation.md:1]: Same as this one, compress the context and put to docs/operators/batch-invariant-logp.md.

File: scripts/compare_logprob.py
Comment (KJLdefeated) [inline scripts/compare_logprob.py:1]: I think we can launch rl_engine/testing/logprob_comparison.py directly in logprob_comparison.py? Since scripts directory contains project-wise scope script, but this is kernel specific.

**Review by KJLdefeated — CHANGES_REQUESTED** [review]
Overall looks good to me. After addressed the request, I am happy to approve.

**Comment by hihaluemen** [comment]
> Overall looks good to me. After addressed the request, I am happy to approve.

@KJLdefeated  Addressed, thanks for the suggestions. I consolidated both harness documents into docs/operators/batch-invariant-logp.md and moved the kernel-specific CLI into rl_engine/testing/logprob_comparison.py.

**Review by KJLdefeated — APPROVED** [review]
